File: app/src/main/java/moe/tlaster/openween/core/api/friendships/Friends.java
package moe.tlaster.openween.core.api.friendships;

import java.util.HashMap;
import java.util.Map;

import moe.tlaster.openween.core.api.Constants;
import moe.tlaster.openween.core.model.user.UserListModel;
import moe.tlaster.openween.core.model.user.UserModel;
import moe.tlaster.openween.common.helpers.HttpHelper;
import moe.tlaster.openween.common.helpers.InvalidAccessTokenException;
import moe.tlaster.openween.common.helpers.JsonCallback;

/**
 * Created by Tlaster on 2016/9/7.
 */
public class Friends {
    private static <T> void getUsers(T uid, int count, int cursor, String url, JsonCallback<UserListModel> callback) {
        Map<String, String> param = new HashMap<>();
        param.put("uid", String.valueOf(uid));
        param.put("count", String.valueOf(count));
        param.put("cursor", String.valueOf(cursor));
        HttpHelper.getAsync(url, param, callback);
    }
    public static void getFriends(long uid, int count, int cursor, JsonCallback<UserListModel> callback) {
        getUsers(uid, count, cursor, Constants.FRIENDSHIPS_FRIENDS, callback);
    }
    public static void getFriends(String screen_name, int count, int cursor, JsonCallback<UserListModel> callback) {
        getUsers(screen_name, count, cursor, Constants.FRIENDSHIPS_FRIENDS, callback);
    }
    public static void getFriendsInCommon(long uid, long suid, int count, int page, JsonCallback<UserListModel> callback) {
        Map<String, String> param = new HashMap<>();
        param.put("uid", String.valueOf(uid));
        param.put("count", String.valueOf(count));
        param.put("page", String.valueOf(page));
        if (suid != -1) param.put("suid", String.valueOf(suid));
        HttpHelper.getAsync(Constants.FRIENDSHIPS_FRIENDS_IN_COMMON, param, callback);
    }
    public static void getBliateral(long uid, int count, int page, int sort, JsonCallback<UserListModel> callback) {
        Map<String, String> param = new HashMap<>();
        param.put("uid", String.valueOf(uid));
        param.put("count", String.valueOf(count));
        param.put("page", String.valueOf(page));
        param.put("sort", String.valueOf(sort));
        HttpHelper.getAsync(Constants.FRIENDSHIPS_FRIENDS_BILATERAL, param, callback);
    }
    public static void getFollowers(long uid, int count, int cursor, JsonCallback<UserListModel> callback) {
        getUsers(uid, count, cursor, Constants.FRIENDSHIPS_FOLLOWERS, callback);
    }
    public static void getFollowers(String screen_name, int count, int cursor, JsonCallback<UserListModel> callback) {
        getUsers(screen_name, count, cursor, Constants.FRIENDSHIPS_FOLLOWERS, callback);
    }
    public static void follow(long uid, JsonCallback<UserModel> callback) {
        Map<String, String> param = new HashMap<>();
        param.put("uid", String.valueOf(uid));
        HttpHelper.postAsync(Constants.FRIENDSHIPS_CREATE, param, callback);
    }
    public static void follow(String screen_name, JsonCallback<UserModel> callback) {
        Map<String, String> param = new HashMap<>();
        param.put("screen_name", String.valueOf(screen_name));
        HttpHelper.postAsync(Constants.FRIENDSHIPS_CREATE, param, callback);
    }
    public static void unfollow(long uid, JsonCallback<UserModel> callback) {
        Map<String, String> param = new HashMap<>();
        param.put("uid", String.valueOf(uid));
        HttpHelper.postAsync(Constants.FRIENDSHIPS_DESTROY, param, callback);
    }
    public static void unfollow(String screen_name, JsonCallback<UserModel> callback) {
        Map<String, String> param = new HashMap<>();
        param.put("screen_name", String.valueOf(screen_name));
        HttpHelper.postAsync(Constants.FRIENDSHIPS_DESTROY, param, callback);
    }
}

File: app/src/main/java/moe/tlaster/openween/core/api/friendships/Friends.kt
package moe.tlaster.openween.core.api.friendships

import android.support.v4.util.ArrayMap

import moe.tlaster.openween.core.api.Constants
import moe.tlaster.openween.core.model.user.UserListModel
import moe.tlaster.openween.core.model.user.UserModel
import moe.tlaster.openween.common.helpers.HttpHelper

/**
 * Created by Tlaster on 2016/9/7.
 */
object Friends {
    private fun <T> getUsers(uid: T, count: Int, cursor: Int, url: String) : UserListModel {
        val param = ArrayMap<String, String>()
        param.put("uid", uid.toString())
        param.put("count", count.toString())
        param.put("cursor", cursor.toString())
        return HttpHelper.getAsync(url, param)
    }

    fun getFriends(uid: Long, count: Int, cursor: Int) : UserListModel {
        return getUsers(uid, count, cursor, Constants.FRIENDSHIPS_FRIENDS)
    }

    fun getFriends(screen_name: String, count: Int, cursor: Int) : UserListModel {
        return getUsers(screen_name, count, cursor, Constants.FRIENDSHIPS_FRIENDS)
    }

    fun getFriendsInCommon(uid: Long, suid: Long, count: Int, page: Int) : UserListModel {
        val param = ArrayMap<String, String>()
        param.put("uid", uid.toString())
        param.put("count", count.toString())
        param.put("page", page.toString())
        if (suid != -1L) param.put("suid", suid.toString())
        return HttpHelper.getAsync(Constants.FRIENDSHIPS_FRIENDS_IN_COMMON, param)
    }

    fun getBliateral(uid: Long, count: Int, page: Int, sort: Int) : UserListModel {
        val param = ArrayMap<String, String>()
        param.put("uid", uid.toString())
        param.put("count", count.toString())
        param.put("page", page.toString())
        param.put("sort", sort.toString())
        return HttpHelper.getAsync(Constants.FRIENDSHIPS_FRIENDS_BILATERAL, param)
    }

    fun getFollowers(uid: Long, count: Int, cursor: Int) : UserListModel {
        return getUsers(uid, count, cursor, Constants.FRIENDSHIPS_FOLLOWERS)
    }

    fun getFollowers(screen_name: String, count: Int, cursor: Int) : UserListModel {
        return getUsers(screen_name, count, cursor, Constants.FRIENDSHIPS_FOLLOWERS)
    }

    fun follow(uid: Long) : UserModel {
        val param = ArrayMap<String, String>()
        param.put("uid", uid.toString())
        return HttpHelper.postAsync(Constants.FRIENDSHIPS_CREATE, param)
    }

    fun follow(screen_name: String) : UserModel {
        val param = ArrayMap<String, String>()
        param.put("screen_name", screen_name)
        return HttpHelper.postAsync(Constants.FRIENDSHIPS_CREATE, param)
    }

    fun unfollow(uid: Long) : UserModel {
        val param = ArrayMap<String, String>()
        param.put("uid", uid.toString())
        return HttpHelper.postAsync(Constants.FRIENDSHIPS_DESTROY, param)
    }

    fun unfollow(screen_name: String) : UserModel {
        val param = ArrayMap<String, String>()
        param.put("screen_name", screen_name)
        return HttpHelper.postAsync(Constants.FRIENDSHIPS_DESTROY, param)
    }
}
